### ISSUE TITLE
Update to Nuxt 4

### ISSUE THREAD

**Issue by andrewtavis** [activist-org/activist#919]
### Terms

- [X] I have searched [open and closed feature requests](https://github.com/activist-org/activist/issues?q=is%3Aissue+label%3Afeature)
- [X] I agree to follow activist's [Code of Conduct](https://github.com/activist-org/activist/blob/main/.github/CODE_OF_CONDUCT.md)

### Description

One thing that the community should get to sooner rather than later is update underlying frameworks early in the process so we don't need to think about this later when the platform grows. The second of these is Nuxt, which will hit version 4.0 soon at the time this issue is being made. This isn't needed for the first release, but should be gotten to by the end of 2024 :)

There's currently no official upgrade guide, but I'll add it to this issue when it's up!

- Blocks #1148

### Contribution

I'll be working on this once the new version of Nuxt is released!

**Comment by CreativeStudio16** [comment]
setting

**Comment by aakashkrsingh1** [comment]
Hey @andrewtavis , a new version released 2 weeks ago. Can I pick this up, please? 
_New to open source._

**Comment by supraaxdd** [comment]
Hi, is this issue still blocked? If not, I'd be able to work on it.

**Comment by andrewtavis** [comment]
Thanks for your interest, @supraaxdd! Looks like Nuxt 4 is out. Could you send along a message before the PR that outlines the changes that need to be made? Will help us to plan reviews and what needs to come in first 😊

**Comment by supraaxdd** [comment]
> Thanks for your interest, [@supraaxdd](https://github.com/supraaxdd)! Looks like Nuxt 4 is out. Could you send along a message before the PR that outlines the changes that need to be made? Will help us to plan reviews and what needs to come in first 😊

I'll take a look at what changes are in the docs vs. what's in the source code and see what needs to be changed. Once I get my head around the whole show, I think I could make a quick summary :)

**Comment by supraaxdd** [comment]
Here's my little update (perhaps a summary) of what I found.

Firstly, I'd like to say that all of the change notes can be found [here](https://nuxt.com/docs/4.x/getting-started/upgrade).

Here is the summary of the changes:
One of the big changes is the directory structure, with the `app` directory being used as the default `srcDir`. 
An excerpt from the migration guide can be found below:
 - the new Nuxt default `srcDir` is `app/` by default, and most things are resolved from there.
  - serverDir now defaults to `<rootDir>/server` rather than `<srcDir>/server`
  - layers/, modules/ and public/ are resolved relative to `<rootDir>` by default
  - if using [Nuxt Content v2.13+](https://github.com/nuxt/content/pull/2649), content/ is resolved relative to `<rootDir>`
  - a new dir.app is added, which is the directory we look for `router.options.ts` and `spa-loading-template.html` - this defaults to `<srcDir>/`

With that in mind, this is what the guide suggests as the steps to migrating:
- Create a new directory called `app/`.
- Move your `assets/`, `components/`, `composables/`, `app/layouts/`, `app/middleware/`, `app/pages/`, `app/plugins/` and `utils/` folders under it, as well as `app.vue`, `error.vue`, `app.config.ts`. If you have an `app/router-options.ts` or `app/spa-loading-template.html`, these paths remain the same.
- Make sure your `nuxt.config.ts`, `content/`, `layers/`, `modules/`, `public/` and `server/` folders remain outside the `app/` folder, in the root of your project.
- Remember to update any third-party configuration files to work with the new directory structure, such as your tailwindcss or eslint configuration (if required - @nuxtjs/tailwindcss should automatically configure tailwindcss correctly).

So for activist, all of the folders as mentioned above, such as `pages/` would be moved into the `app/` directory.

If required, there is also a way to force nuxt to continue using v3 convention:
```js
export default defineNuxtConfig({
  // This reverts the new srcDir default from `app` back to your root directory
  srcDir: '.',
  // This specifies the directory prefix for `router.options.ts` and `spa-loading-template.html`
  dir: {
    app: 'app'
  }
})

```
although since we are migrating, I doubt this would be of much use.

Another key thing that's changed is the [Data Fetching Layer](https://nuxt.com/docs/4.x/getting-started/upgrade#singleton-data-fetching-layer). When using `useFetch` or `useAsyncData`, the refs to `data`, `status`, and `error` are shared if the key is the same. Upon a closer look at the code, I don't think it will be an issue because any data fetching operations are in their own functions, therefore in their own local scope (unless I'm wrong, feel free to correct me ;))

Another change related to the one outlined can be found [here](https://nuxt.com/docs/4.x/getting-started/upgrade#shared-prerender-data). Furthermore, reactive key behaviour has changed whereby `useFetch` and `useAsyncData` will automatically refetch the data, unless provided with the `immediate: false` flag. See more about this [here](https://nuxt.com/docs/4.x/getting-started/upgrade#key-change-behavior-in-useasyncdata-and-usefetch)

That is the only significant change in terms of a major rewrite that I spotted. A lot of the changes involving these two functions are in the backend, such as data cleaning to lessen memory usage and "reactive key support".

Another change of lesser importance, such as module loading order don't require changes, although if the application breaks, it could be a problematic area if activist relies on the module loading order. I'm not entirely sure if there is anything there, but keeping an eye on that would be beneficial.

The loading spa screen has also slightly changed, with the loading screen itself moving outside of the root layout. This is a minor change, although it will yield a better UI as there will be less white screen flashing. See [here](https://nuxt.com/docs/4.x/getting-started/upgrade#new-dom-location-for-spa-loading-screen)

Currently the error.vue screen doesn't use error.data, however, going forward, this property will no longer need to be parsed (using JSON.parse) as it comes parsed with the error object. See [here](https://nuxt.com/docs/4.x/getting-started/upgrade#parsed-errordata)

There will be no more inline styles for global CSS, to improve download size of the webpage. See [here](https://nuxt.com/docs/4.x/getting-started/upgrade#more-granular-inline-styles)

Page Metadata will be scanned *after* calling pages:extend. This is available in the nuxt.config.ts 
https://github.com/activist-org/activist/blob/0a3a89e3dbc5723f0317a6bf49591d6a98f4621a/frontend/nuxt.config.ts#L138-L145

The change is minor and simple to do, outlined [here](https://nuxt.com/docs/4.x/getting-started/upgrade#scan-page-meta-after-resolution)

The rest of the changes, don't really affect the project from what I can tell, but a deeper assessment of the codebase would more than likely reveal far more than what I looked into. 

This is my summary from the quick perusal around the codebase. This is my first contribution so I'm still relatively new to the flow of things, but I'm happy to learn how it's done!

@andrewtavis

**Comment by andrewtavis** [comment]
Thanks for outlining the above, @supraaxdd! @nicki182, this is obviously a massive change. Do you have a suggestion for when we should do this? Might make sense before the release, but we should definitely get the resources work finished, and maybe also finish the frontend dependency updates too so that it's not such a big jump and most of the TS errors we could expect would be finished before?

**Comment by supraaxdd** [comment]
> Thanks for outlining the above, @supraaxdd! @nicki182, this is obviously a massive change. Do you have a suggestion for when we should do this? Might make sense before the release, but we should definitely get the resources work finished, and maybe also finish the frontend dependency updates too so that it's not such a big jump and most of the TS errors we could expect would be finished before?

@andrewtavis If there's anything I could help with in the meantime, I'm happy to pick something up regarding the resource changes mentioned. I could try dive in and see what's happening? Could you send the issue/PR with the changes regarding the resources?

**Comment by nicki182** [comment]
The resources is almost done, but you could help with the e2e testing for organization pages. #1467

**Comment by andrewtavis** [comment]
@supraaxdd, one thing that's not quite as glorious but could be done is working through some of the dependency updates for the frontend. We currently have a ton of [PRs related to this](https://github.com/activist-org/activist/pulls). We always do explicit dependencies for the frontend. Let us know if you'd like to look into this as well!

**Comment by supraaxdd** [comment]
I can give it a shot, I'm currently doing this in the side because of work but I'll certainly take a look at it in the coming days, perhaps even tomorrow! I'll keep you posted on progress throughout.

**Comment by supraaxdd** [comment]
@andrewtavis @nicki182 since there is other work to be done, is this issue blocked by those issues/PRs?

**Comment by nicki182** [comment]
We first need to make sure it's stable before implementing this massive change. I think it should be done once we are in MVP state. Since the release of nuxt 4 as stable release was in July of this year, so we are not in a hurry. If a year happened then yes. @andrewtavis what do you think?

**Comment by nicki182** [comment]
@supraaxdd it's not necessarily blocked but it should be given lower priority since they are other issues that should be given bigger priority

**Comment by andrewtavis** [comment]
I'd say we should get this done before the release just so we don't need to react to it once we're live, but agree that for right now there are lots of things that we should do beforehand. Maybe you can work on some dependency work for a moment to set the stage for this, and then we can move towards it once we're finished with the work we're doing on the resource components and authentication?

**Comment by nicki182** [comment]
Maybe, next week dev sync since it's a dev night we can see about doing this at once.

**Comment by andrewtavis** [comment]
Ya that would make sense :)

**Comment by supraaxdd** [comment]
Quick update on my part...

I was trying to get the playwright tests to work with yarn for the other open issues that we have, but I'm having issues getting corepack specifically to behave itself. I might need a bit more time, but I'd say that eventually I'll get there

**Comment by andrewtavis** [comment]
All good, @supraaxdd! Please let us know if the [Development channel on Matrix](https://matrix.to/#/!CRgLpGeOBNwxYCtqmK:matrix.org?via=matrix.org&via=systemli.org&via=librezo.fr) if you're experiencing bugs that you can't figure out :)

**Comment by andrewtavis** [comment]
Closed by #1547 🚀 Thanks for the great work getting work getting through all this with me, @nicki182!
